src/relax/transform/lift_transform_params.cc

```diff
@@ -37,405 +38,467 @@
+    return func;
+  }
 
-    // Create mapping from the original input variables to the TupleGetItem from the packed
-    // parameter tuple Add the parameters that are marked as the output of the function to the
-    // output tuple
-    for (const auto& input : inputs_) {
-      input_remap_.emplace(input.get(), TupleGetItem(params, input_remap_.size()));
-      if (outputs_.count(input)) {
-        auto output_var = builder_->Emit(input_remap_.at(input.get()));
-        f_add_output(input, output_var);
-      }
+  Function MakeRuntimeFunction() const {
+    Array<Binding> bindings;
+
+    // Any parameter that isn't available until runtime must be an
+    // input, along with any output from the compile-time function.
+    // Compile-time outputs must have a fresh non-dataflow var to
+    // serve as the parameter.  This trivial binding will later be
+    // removed with CanonicalizeBindings.
+    Array<Var> params = GetRuntimeInputs();
+    if (auto propagated_tir_vars = GetPropagatedSymbolicVariables(); propagated_tir_vars.size()) {
+      ShapeStructInfo shape_sinfo(
+          propagated_tir_vars.Map([](tir::Var var) -> PrimExpr { return var; }));
+      Var shape_expr("vars_from_compile_time_params", shape_sinfo);
+      params.push_back(shape_expr);
+    }
+    for (const auto& var : GetCompileTimeOutputs()) {
+      Var param_var(var->name_hint(), GetStructInfo(var));
+      bindings.push_back(VarBinding(var, param_var));
+      params.push_back(param_var);
     }
 
-    // Re-emit the bindings that are lifted. Update the output tuple if the binding is marked as the
-    // output.
-    for (const auto& binding : bindings_) {
-      if (outputs_.count(binding->var)) {
-        auto output_var = builder_->Emit(VisitExpr(binding->value));
-        var_remap_[binding->var->vid] = output_var;
-        f_add_output(binding->var, output_var);
-      } else {
-        VisitBinding(binding);
+    // Any binding that is computable at compile-time should be
+    // suppressed at run-time.
+    std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> to_suppress;
+    for (const auto& binding : computable_at_compile_time) {
+      if (requires_compile_time_param.count(binding->var)) {
+        to_suppress.insert(binding->var);
       }
     }
 
-    // Create the function.
-    Expr transformed_params = builder_->EmitOutput(Tuple(output_vars));
-    BindingBlock block = builder_->EndBlock();
-    Expr body = VisitWithNewScope(SeqExpr({block}, transformed_params), Array<Var>{params});
-    Function f_transform_params =
-        Function(/*params=*/{params}, /*body=*/body, /*ret_struct_info=*/NullOpt);
-    return {f_transform_params, output_to_index};
-  }
+    struct SuppressCompileTime : ExprMutator {
+      std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> to_suppress;
+      explicit SuppressCompileTime(
+          std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> to_suppress)
+          : to_suppress(to_suppress) {}
```

Comment:
Question separate from the review: Are we going with the convention of having member var names end in an underscore? I'm not a partisan on that one, but we should be consistent. @tqchen

Reply:
The TVM convention, as I understand it, is to have private members have a trailing underscore, while public members do not.  Often if I'm making quick subclasses of `ExprMutator`, I'll have a struct with public fields rather than a class with private fields.  Since the class definition already occurs within a function scope, the visibility of the entire class is already restricted, and additional visibility restrictions of private members isn't necessary.

That said, I've updated it to be a class with a private `to_suppress_`.  I figure that if a deviation from convention is big enough to raise a question, it's big enough to avoid the deviation altogether.

Reply:
I was just curious, I don't think it's worthy of holding up a review or anything like that. Having a different rule for public/private members is reasonable.